### PR TITLE
Update project key

### PR DESCRIPTION
**Pull request by teryk**

### CODE REVIEW

File: server/sonar-server/src/main/java/org/sonar/server/component/ws/BulkUpdateKeyAction.java [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+import org.sonar.api.server.ws.Response;
+import org.sonar.api.server.ws.WebService;
+import org.sonar.db.DbClient;
+import org.sonar.db.DbSession;
+import org.sonar.db.component.ComponentDto;
+import org.sonar.server.component.ComponentFinder;
+import org.sonar.server.component.ComponentFinder.ParamNames;
+import org.sonar.server.component.ComponentService;
+import org.sonarqube.ws.client.component.BulkUpdateWsRequest;
+import org.sonarqube.ws.client.component.UpdateWsRequest;
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/component/ws/BulkUpdateKeyAction.java]: MINOR Remove this unused import 'org.sonarqube.ws.client.component.UpdateWsRequest'. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AUselessImportCheck)

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue
* MINOR 1 minor

Watch the comments in this conversation to review them.

File: server/sonar-server/src/main/java/org/sonar/server/component/ws/BulkUpdateKeyAction.java [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+import org.sonar.api.server.ws.Response;
+import org.sonar.api.server.ws.WebService;
+import org.sonar.db.DbClient;
+import org.sonar.db.DbSession;
+import org.sonar.db.component.ComponentDto;
+import org.sonar.server.component.ComponentFinder;
+import org.sonar.server.component.ComponentFinder.ParamNames;
+import org.sonar.server.component.ComponentService;
+import org.sonarqube.ws.client.component.BulkUpdateWsRequest;
+
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/component/ws/BulkUpdateKeyAction.java]: MINOR Remove this unused import 'org.sonarqube.ws.client.component.UpdateWsRequest'. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AUselessImportCheck)

File: sonar-db/src/main/java/org/sonar/db/component/ResourceKeyUpdaterDao.java [hunk 3]
```diff
@@ -19,6 +19,7 @@
  */
 package org.sonar.db.component;
 
+import com.google.common.base.Preconditions;
```
Comment (SonarTech) [inline sonar-db/src/main/java/org/sonar/db/component/ResourceKeyUpdaterDao.java]: MINOR Remove this unused import 'com.google.common.base.Preconditions'. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AUselessImportCheck)

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 issues
* MINOR 2 minor

Watch the comments in this conversation to review them.

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue
* MINOR 1 minor

Watch the comments in this conversation to review them.

File: sonar-db/src/main/java/org/sonar/db/component/ResourceKeyUpdaterDao.java [hunk 4]
```diff
@@ -19,6 +19,7 @@
  */
```
Comment (simonbrandhof) [inline sonar-db/src/main/java/org/sonar/db/component/ResourceKeyUpdaterDao.java]: This class could be renamed to ComponentKeyUpdaterDao at the same time.

File: server/sonar-server/src/main/java/org/sonar/server/component/ws/BulkUpdateKeyAction.java [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+    action.createParam(PARAM_ID)
+      .setDescription("Project or module id")
+      .setExampleValue(UUID_EXAMPLE_01);
+
+    action.createParam(PARAM_KEY)
+      .setDescription("Project or module key")
+      .setExampleValue("my_old_project");
+
+    action.createParam(PARAM_FROM)
```
Comment (simonbrandhof) [inline server/sonar-server/src/main/java/org/sonar/server/component/ws/BulkUpdateKeyAction.java]: I don't understand the 2 parameters "from" and "to". It deserves probably better example and naming.

File: sonar-db/src/main/java/org/sonar/db/component/ComponentKeyUpdaterDao.java [hunk 6]
```diff
@@ -137,13 +161,17 @@ private static void runBatchUpdateForAllResources(Collection<ResourceDto> resour
...
-        throw new IllegalStateException("Impossible to update key: a resource with \"" + newName + "\" key already exists.");
+      String newKey = computeNewKey(module, stringToReplace, replacementString);
+      checkArgument(isValidModuleKey(newKey), "Malformed key for '%s'. Allowed characters are alphanumeric, '-', '_', '.' and ':', with at least one non-digit.", newKey);
+      if (mapper.countResourceByKey(newKey) > 0) {
+        throw new IllegalArgumentException("Impossible to update key: a component with key \"" + newKey + "\" already exists.");
       }
     }
   }
 
+  private ComponentKeyUpdaterMapper mapper(DbSession dbSession) {
```
Comment (SonarTech) [inline sonar-db/src/main/java/org/sonar/db/component/ComponentKeyUpdaterDao.java]: MAJOR Make "mapper" a "static" method. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS2325)

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 issues
* MAJOR 2 major

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. MAJOR [ComponentKeyUpdaterDao.java#L125](https://github.com/SonarSource/sonarqube/blob/1437c83d435a572d2ff8e31e0d3fb9befbc0bb7d/sonar-db/src/main/java/org/sonar/db/component/ComponentKeyUpdaterDao.java#L125): Combine this loop with the one that starts on line 120. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS3047)

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 issues
* MAJOR 2 major

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. MAJOR [ComponentKeyUpdaterDao.java#L133](https://github.com/SonarSource/sonarqube/blob/64a9766460443dfcc37474d45fdb98159a33c56b/sonar-db/src/main/java/org/sonar/db/component/ComponentKeyUpdaterDao.java#L133): Combine this loop with the one that starts on line 128. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS3047)

**Pull request closed without merging**